src/FirstClaim/FirstClaim.sol

```diff
@@ -0,0 +1,214 @@
+
+    ////////////////////////////////////////////////////////////////
+    ///                      CONFIG STORAGE                      ///
+    ////////////////////////////////////////////////////////////////
+
+    /// @notice address of the Safe Allowance Module
+    AllowanceModule public ALLOWANCE_MODULE;
+
+    /// @notice Worldcoin token address
+    address public WLD_TOKEN;
+
+    /// @notice BVI Safe that grants allowances to this contract
+    GnosisSafe public HOLDER;
+
+    /// @notice Recurring Grant Drop contract
+    IRecurringGrantDrop public RECURRING_GRANT_DROP;
+
+    /// @notice addresses that can call the claim function
+    mapping(address => bool) public allowedCallers;
+
+    ////////////////////////////////////////////////////////////////
+    ///                       CONSTRUCTOR                        ///
+    ////////////////////////////////////////////////////////////////
+
+    constructor(
+        address _allowanceModuleAddress,
+        address _wldToken,
+        address _holder,
+        address _recurringGrantDrop
+    ) Ownable(msg.sender) {
+        if (
+            _allowanceModuleAddress == address(0) || _wldToken == address(0)
+                || _holder == address(0) || _recurringGrantDrop == address(0)
+        ) {
+            revert ZeroAddress();
+        }
+        ALLOWANCE_MODULE = AllowanceModule(_allowanceModuleAddress);
+        WLD_TOKEN = _wldToken;
+        HOLDER = GnosisSafe(_holder);
+        RECURRING_GRANT_DROP = IRecurringGrantDrop(_recurringGrantDrop);
+    }
+
+    ////////////////////////////////////////////////////////////////
+    ///                        FUNCTIONS                         ///
+    ////////////////////////////////////////////////////////////////
+
+    /// @notice Claim the first grant
+    /// @param grantId The grant ID to claim
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+    /// @param amount The total amount to claim
+    function claim(
+        uint256 grantId,
+        address receiver,
+        uint256 root,
+        uint256 nullifierHash,
+        uint256[8] calldata proof,
+        uint256 amount
+    ) external {
+      if (!allowedCallers[msg.sender]) {
+        revert OnlyAllowedCaller();
+      }
+
+      uint256 currentGrantAmount = RECURRING_GRANT_DROP.grant().getAmount(grantId);
+      if (currentGrantAmount >= amount) {
+        revert GrantAmountTooLarge();
+      }
+
+      RECURRING_GRANT_DROP.claim(grantId, receiver, root, nullifierHash, proof);
+      ALLOWANCE_MODULE.executeAllowanceTransfer(
+        HOLDER,
+        WLD_TOKEN,
+        payable(receiver),
+        uint96(amount - currentGrantAmount)
```

Comment:
do you maybe want to have some safety limit on the amount at least? one backend bug empties the wallet